File: algorithms/swift/278-first-bad-version.swift
// https://leetcode.com/problems/first-bad-version/

/**
 * The knows API is defined in the parent class VersionControl.
 *     func isBadVersion(_ version: Int) -> Bool{}
 */

class Solution : VersionControl {
    func firstBadVersion(_ n: Int) -> Int {
        if (isBadVersion(1)) {
            return 1
        }
        var mid = Int()
        var l = 2, r = n
        while(l <= r) {
            mid = (l+r)/2
            if (isBadVersion(mid) && !isBadVersion(mid-1)) {
                return mid
            }
            if (isBadVersion(mid)) {
                r = mid-1
            } else {
                l = mid+1
            }
        }
        return -1
    }
}

File: algorithms/c++/278-first-bad-version.cpp
// https://leetcode.com/problems/first-bad-version/
#include <bits/stdc++.h>

using namespace std;

// The API isBadVersion is defined for you.
// bool isBadVersion(int version);

class Solution {
public:
    int firstBadVersion(int n) {
        int left = 1;
        int right = n;
        
        while(left <= right) {
            int mid = left + (right - left)/2; // bypass int overflow
            if (isBadVersion(mid) && (mid == 1 || !isBadVersion(mid-1)))
                return mid;
            if (isBadVersion(mid))
                right = mid - 1;
            else
                left = mid + 1;
        }
        return -1;
    }
};

int main() {

    return 0;
}